File: abilities-common/src/main/java/io/github/solusmods/eternalcore/abilities/impl/network/c2s/RequestAbilityActivationPacket.java
package io.github.solusmods.eternalcore.abilities.impl.network.c2s;

import dev.architectury.networking.NetworkManager;
import dev.architectury.utils.Env;
import io.github.solusmods.eternalcore.abilities.EternalCoreAbilities;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;

public record RequestAbilityActivationPacket(int keyNumber,
                                             ResourceLocation abilityId,
                                             int mode) implements CustomPacketPayload {

    public static final CustomPacketPayload.Type<RequestAbilityActivationPacket> TYPE = new CustomPacketPayload.Type<>(EternalCoreAbilities.create("request_ability_activation"));
    public static final StreamCodec<FriendlyByteBuf, RequestAbilityActivationPacket> STREAM_CODEC = CustomPacketPayload.codec(RequestAbilityActivationPacket::encode, RequestAbilityActivationPacket::new);

    public RequestAbilityActivationPacket(FriendlyByteBuf buf) {
        this(buf.readInt(), buf.readResourceLocation(), buf.readInt());
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeInt(this.keyNumber);
        buf.writeResourceLocation(this.abilityId);
        buf.writeInt(this.mode);
    }

    public void handle(NetworkManager.PacketContext context) {
        if (context.getEnvironment() != Env.SERVER) return;
        context.queue(() -> ClientAccess.handle(this, context.getPlayer()));
    }



    public @NotNull Type<RequestAbilityActivationPacket> type() {
        return TYPE;
    }
}

File: abilities-common/src/main/kotlin/io/github/solusmods/eternalcore/abilities/impl/network/c2s/RequestAbilityActivationPacket.kt
package io.github.solusmods.eternalcore.abilities.impl.network.c2s

import dev.architectury.networking.NetworkManager
import dev.architectury.utils.Env
import io.github.solusmods.eternalcore.abilities.EternalCoreAbilities
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamDecoder
import net.minecraft.network.codec.StreamMemberEncoder
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation

@JvmRecord
data class RequestAbilityActivationPacket(
    val keyNumber: Int,
    val abilityId: ResourceLocation?,
    val mode: Int
) : CustomPacketPayload {
    constructor(buf: FriendlyByteBuf) : this(buf.readInt(), buf.readResourceLocation(), buf.readInt())

    fun encode(buf: FriendlyByteBuf) {
        buf.writeInt(this.keyNumber)
        buf.writeResourceLocation(this.abilityId)
        buf.writeInt(this.mode)
    }

    fun handle(context: NetworkManager.PacketContext) {
        if (context.getEnvironment() != Env.SERVER) return
        context.queue(Runnable { ClientAccess.handle(this, context.getPlayer()) })
    }


    override fun type(): CustomPacketPayload.Type<RequestAbilityActivationPacket?> {
        return TYPE
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<RequestAbilityActivationPacket?> =
            CustomPacketPayload.Type<RequestAbilityActivationPacket?>(EternalCoreAbilities.create("request_ability_activation"))
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf?, RequestAbilityActivationPacket?> =
            CustomPacketPayload.codec<FriendlyByteBuf?, RequestAbilityActivationPacket?>(
                StreamMemberEncoder { obj: RequestAbilityActivationPacket?, buf: FriendlyByteBuf? -> obj!!.encode(buf!!) },
                StreamDecoder { buf: FriendlyByteBuf? ->
                    RequestAbilityActivationPacket(
                        buf!!
                    )
                })
    }
}
